Слушатель Kafka-эвентов для комментов

Kot-Mars opened the pull request:
Jira - https://faang-school.atlassian.net/browse/BJS2-66672

Commit:
BJS2-66672: create KafkaCommentConsumer

Commit:
BJS2-66672: create KafkaCommentConsumer
Create RedisCommentServiceImpl, unit-test

s3nkyh commented on src/main/java/faang/school/postservice/config/kafka/KafkaCommentConsumer.java:
```diff
@@ -0,0 +1,24 @@
+
+    @KafkaListener(topics = "${spring.kafka.listener.topics.news-feed-comment}",
+            groupId = "${spring.kafka.listener.groupId.news-feed-group}")
+    public void listen(CommentEvent event, Acknowledgment ack) {
```
Ты ловишь сразу эвент, хотя должен придти json который нужно заммапить в эвент

s3nkyh commented on src/main/java/faang/school/postservice/service/redis/RedisCommentServiceImpl.java:
```diff
@@ -0,0 +1,45 @@
+        String redisKey = String.format(redisKeyPrefix, event.getPostId());
+        double score = (double) Instant.now().toEpochMilli();
+        try {
+            String json = mapper.writeValueAsString(event);
```
Можешь если хочешь написать свой JsonUtils и будет удобнее и меньше кода. Можешь посмотреть у меня

s3nkyh commented on src/main/java/faang/school/postservice/service/redis/RedisCommentServiceImpl.java:
```diff
@@ -0,0 +1,45 @@
+        double score = (double) Instant.now().toEpochMilli();
+        try {
+            String json = mapper.writeValueAsString(event);
+            redisTemplate.opsForZSet().add(redisKey, json, score);
```
может сначала проверять размер в кэше а потом добавлять

s3nkyh commented on src/main/java/faang/school/postservice/config/kafka/KafkaCommentConsumer.java:
```diff
@@ -0,0 +1,24 @@
+package faang.school.postservice.config.kafka;
```
слушатель в кофниге тоже странно, либо я туплю

Kot-Mars committed:
BJS2-66672: Исправление ошибок
Перенес kafka из конфига в отдельную папку
Добавление JsonUtils
Исправление KafkaCommentConsumer

PavelEzhkov commented on src/main/java/faang/school/postservice/kafka/KafkaCommentConsumer.java:
```diff
@@ -0,0 +1,35 @@
+            redisCommentService.saveComment(event);
+            log.info("Received message {}", event);
+            ack.acknowledge();
+        } catch (Exception e) {
```
лучше ловить бы ловить более узкие ошибки

PavelEzhkov commented on src/main/java/faang/school/postservice/kafka/KafkaCommentConsumer.java:
```diff
@@ -0,0 +1,35 @@
+            ack.acknowledge();
+        } catch (Exception e) {
+            log.error("Failed to process Kafka message: {}", message, e);
+            throw new KafkaConsumerException(String.format(FAILED_TO_DESERIALIZE_EVENT, CommentEvent.class.getSimpleName()));
```
В текущей реализации при ошибке десериализации выбрасывается исключение, но не ясно, как это влияет на обработку сообщения в Kafka,можно попасть в цикл если будет поломанное сообщение

PavelEzhkov commented on src/main/java/faang/school/postservice/service/redis/RedisCommentServiceImpl.java:
```diff
@@ -0,0 +1,45 @@
+@RequiredArgsConstructor
+public class RedisCommentServiceImpl implements RedisCommentService {
+    private final RedisTemplate<String, String> redisTemplate;
+    private final String redisKeyPrefix = "post:%d:comment";
```
конечно и это лучше вынести в проперти

PavelEzhkov commented on src/main/java/faang/school/postservice/service/redis/RedisCommentServiceImpl.java:
```diff
@@ -0,0 +1,45 @@
+        String redisKey = String.format(redisKeyPrefix, event.getPostId());
+        double score = (double) Instant.now().toEpochMilli();
+        try {
+            Long size = redisTemplate.opsForZSet().zCard(redisKey);
```
Текущая реализация содержит race condition между проверкой размера и удалением старых элементов

Kot-Mars committed:
BJS2-66672: fixCode

Kot-Mars committed:
Merge branch 'medusa-master-stream9' into BJS2-66672

# Conflicts:
#	src/main/resources/application.yaml

Kot-Mars committed:
Merge branch 'medusa-master-stream9' into BJS2-66672

# Conflicts:
#	src/main/java/faang/school/postservice/contants/ErrorMessage.java
#	src/main/resources/application.yaml
